File: LeetCode/BinaryPreOrder.py
# Ques :- 144 ... Pre-order Traversal of binary tree.
class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution(object):
    def preorderTraversal(self, root):
        if root is None:
            return []

        result = []

        def dfs(node):
            if node is None:
                return
            result.append(node.val)
            dfs(node.left)
            dfs(node.right)

        dfs(root)
        return result


# Example usage:
root = TreeNode(1, TreeNode(2, TreeNode(4), TreeNode(5)), TreeNode(3))
print(Solution().preorderTraversal(root))
# Output: [1, 2, 4, 5, 3]

File: BasicToAdvance/Java/LeetCode/BinaryPreorder.java
public class BinaryPreorder {
    static class TreeNode {
        int val;
        TreeNode right;
        TreeNode left;

        TreeNode(int val) {
            this.val = val;
        }
    }

    static void preOrder(TreeNode node, java.util.List<Integer> result) {
        if (node == null) {
            return;
        }

        result.add(node.val);
        preOrder(node.left, result);
        preOrder(node.right, result);
    }

    static int[] Binary_PreOrder(TreeNode root) {
        java.util.List<Integer> result = new java.util.ArrayList<>();
        preOrder(root, result);

        int[] ans = new int[result.size()];
        for (int i = 0; i < result.size(); i++) {
            ans[i] = result.get(i);
        }
        return ans;
    }

    public static void main(String[] args) {
        TreeNode root = new TreeNode(1);
        root.left = new TreeNode(2);
        root.right = new TreeNode(3);
        root.left.left = new TreeNode(4);
        root.left.right = new TreeNode(5);

        int[] ans = Binary_PreOrder(root);
        for (int value : ans) {
            System.out.print(value + " ");
        }
    }
}
